// piles.io/server.js
(function() {
  var auth, db, settings, sys, utils;
  sys = require('sys');
  db = require('piles.io/db');
  utils = require('piles.io/utils');
  settings = (require('piles.io/settings')).settings;
  auth = require('piles.io/auth');
  exports.register = function(app) {
    app.get('/login', function(req, res) {
      return res.render('login', {
        email: 'Email'
      });
    });
    app.post('/login', function(req, res) {
      var email, hashedPwd;
      if (!req.body.email || !req.body.password) {
        return res.render('login', {
          email: req.body.email,
          errors: ['No username or password']
        });
      }
      hashedPwd = utils.hashPassword(req.body.password);
      email = req.body.email.toLowerCase();
      return db.collection('users', function(users) {
        return users.findOne({
          email: email,
          password: hashedPwd
        }, function(err, user) {
          if (!user) {
            return res.render('login', {
              email: email,
              errors: ['Bad email or password']
            });
          }
          auth.doLogin(req, user);
          return db.collection('piles', function(piles) {
            return (piles.find({
              emails: email
            })).toArray(function(err, userPiles) {
              if (userPiles) {
                return res.redirect('/' + userPiles[0]['name']);
              } else {
                return res.redirect('/broke');
              }
            });
          });
        });
      });
    });
    return app.get('/:pilename', function(req, res) {
      return db.collection('piles', function(piles) {
        return piles.findOne({
          name: req.params.pilename
        }, function(err, pile) {
          if (pile) {
            return res.redirect('http://' + settings.CONTENT_DOMAIN + '/app#' + req.params.pilename);
          } else {
            return res.send(404);
          }
        });
      });
    });
  };
}).call(this);
